### PR TITLE
EWPP-1749: Cache tags for vocabulary associations on json api.

### PR DESCRIPTION
**Pull request by yenyasinn**

### CODE REVIEW

**Review by nagyad — CHANGES_REQUESTED** [review]
We need to find a way to cover this code with tests.

File: tests/src/Kernel/EntityAccessTest.php [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+use Drupal\entity_test\Entity\EntityTest;
+use Drupal\field\Entity\FieldConfig;
+use Drupal\field\Entity\FieldStorageConfig;
+use Drupal\open_vocabularies\OpenVocabularyAssociationInterface;
+use Drupal\KernelTests\KernelTestBase;
+use Drupal\Tests\open_vocabularies\Traits\VocabularyTestTrait;
+use Drupal\Tests\user\Traits\UserCreationTrait;
+
+/**
+ * Tests the computed field item list class.
```
Comment (nagyad) [inline tests/src/Kernel/EntityAccessTest.php]: Fix this comment.

File: open_vocabularies.module [hunk 2]
```diff
@@ -56,3 +59,44 @@ function open_vocabularies_form_alter(&$form, FormStateInterface $form_state, $f
...
+  }
+
+  $associations = \Drupal::entityTypeManager()->getStorage('open_vocabulary_association')->loadMultiple();
+  foreach ($associations as $association) {
+    // Vocabulary association uses the field from the entity - add
+    // association as the cache dependency.
+    if (!empty(array_intersect($fields, $association->get('fields')))) {
+      $access->addCacheableDependency($association);
+    }
+  }
```
Comment (nagyad) [inline open_vocabularies.module]: Let's document why we need to this like this.

**Review by nagyad — CHANGES_REQUESTED** [review]

**Review by nagyad — DISMISSED** [review]

**Review by imanoleguskiza — DISMISSED** [review]

**Review by brummbar — CHANGES_REQUESTED** [review]
Queueing myself for review.

File: src/OpenVocabularyAssociationStorage.php [hunk 3]
```diff
@@ -160,6 +161,18 @@ protected function doPostSave(EntityInterface $entity, $update) {
...
 
+    // Invalidate cache for all entity types affected.
+    // We need to do this to guarantee entity lists are
+    // correctly expired when associations are removed.
+    foreach ($entities as $entity) {
+      $fields = $entity->getFields();
+      foreach ($fields as $field) {
+        [$entity_type_id, $bundle, $field_id] = explode('.', $field);
+        $entity_type = \Drupal::entityTypeManager()->getDefinition($entity_type_id);
+        Cache::invalidateTags($entity_type->getListCacheTags());
```
Comment (yenyasinn) [inline src/OpenVocabularyAssociationStorage.php:172]: Don't we need to update `/node/nid` and `/jsonapi/node/[bundle]/[uid]` pages after removing vocabulary associations? These pages don't contain list tags

**Comment by hernani** [comment]
Should be closed

**Pull request closed without merging**